### ISSUE TITLE
No sound in Windows 10 xpra client

### ISSUE THREAD

**Issue by svargh** [Xpra-org/xpra#3053]
Sound from native Windows firefox is working, but not from xpra client.

![no_sound](https://user-images.githubusercontent.com/13705625/111037465-5d10c000-8424-11eb-92ae-ce5d0e010dea.png)

![sessioninfo](https://user-images.githubusercontent.com/13705625/111037702-a1509000-8425-11eb-9a70-d5ec6e3679bd.png)

[serverlog.txt](https://github.com/Xpra-org/xpra/files/6135156/serverlog.txt)

[no-sound.txt](https://github.com/Xpra-org/xpra/files/6135152/no-sound.txt)

**Comment by totaam** [comment]
Information gleaned from the screenshots and logs:
* Ubuntu 20.04 focal server
* Windows 10 aero client version 4.1-r28562 64-bit
* This client system seems to be a virtual machine: `SVGA3D..`

The session info window is showing the audio buffer is receiving and processing data as expected.

Does the problem occur with other applications than Firefox?
Does the problem occur with non-virtualized OS?

Can you run the client from the command line:
```
Xpra_cmd.exe attach tcp://HOST:PORT/ -d sound,gstreamer
```
And post the output?
Perhaps try a different codec too: `--speaker-codec=vorbis` or even `--speaker-codec=mp3`.

Finally, you may want to try a different gstreamer audio sink:
```
Xpra_cmd.exe --env=XPRA_SOUND_SINK=autoaudiosink
```
You can also try: `wasapisink` and `directsoundsink`.

**Comment by svargh** [comment]
I managed to get sound on all clients, once I clicked on the check mark ("Set as fallback")
of 'Xpra Microphone' in tab Output Devices in pavucontrol. 
All clients behave same.
Also I can hear now the slider change tone in pavucontrol.
But it is using the microphone, is there a way to use the "Xpra Speaker"?
Sound with microphone is not so useful,
it is listed in firefox(https://www.onlinemictest.com/), but it catches
up the played sound:
![xpra-issue-1 2-soundwith_microphone](https://user-images.githubusercontent.com/13705625/111064800-73675c00-84b6-11eb-8423-603dbf954ff4.png)
![xpra-issue-1 1-toggle](https://user-images.githubusercontent.com/13705625/111064804-76fae300-84b6-11eb-8719-e503aeea64bd.png)
The connection command was: 
`Xpra_cmd.exe attach tcp://vmuser@192.168.90.242:10000/`
Seems to be a server problem?



**Testsetup:**
- I disabled Ubuntu desktop (service gdm disable), rebooted to console, killed pulseaudio instances
- Started xpra server:
  xpra start --start=gnome-terminal --bind-tcp=0.0.0.0:10000 --pulseaudio=yes --opengl=yes --webcam=no --speaker=on --microphone=off --sharing=yes --html=on --daemon=no

- Xpra client tested on Windows Guest, Windows Host, HtmlClient in Guest Firefox, HtmlClient in Host Firefox.
  All clients have same behaviour:
  Windows Clients: C:\Program Files\Xpra>Xpra_cmd.exe attach tcp://vmuser@192.168.90.242:10000/
  Minor Issue: Only on Windows Host, I always have to toggle once speaker on/off, then sound comes.
  In Html Client: I changed Audio to mp3, then sound works, too - in host - and guest firefox
  
  When trying to click on checkmark of "Xpra Speaker", all clients are silent again, 
  although I can see some sound traffic is coming


The following answers are performed  with Xpra Loudspeaker - not changing to microphone:
**Does the problem occur with other applications than Firefox?**
Yes, when changing slider in pavucontrol, then normally a change tone must come -> Not coming.

**Does the problem occur with non-virtualized OS?**
Yes, I tried the client on Host System, Tried it on HTML Client.


**Can you run the client from the command line: Xpra_cmd.exe attach tcp://HOST:PORT/ -d sound,gstreamer**

Here a log without changes in pavucontrol (Xtra Speaker as output) - NO Sound:
[1clientconnect1_server_speaker_nosound.txt](https://github.com/Xpra-org/xpra/files/6136431/1clientconnect1_server_speaker_nosound.txt)
[1log_server1_speaker_nosound.txt](https://github.com/Xpra-org/xpra/files/6136432/1log_server1_speaker_nosound.txt)

Here a log changes in pavucontrol (Xtra Microphone as output) - WORKING Sound:
[2clientconnect2_server_mic_hassound.txt](https://github.com/Xpra-org/xpra/files/6136433/2clientconnect2_server_mic_hassound.txt)
[2log_server2_mic_hassound.txt](https://github.com/Xpra-org/xpra/files/6136434/2log_server2_mic_hassound.txt)


**Changing the sink and speaker codec hasn't helped either.**
I tried these combinations with the server args in Testsetup:
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=autoaudiosink
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=directsoundsink
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=wasapisink 

**--speaker-codec=mp3:**
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=autoaudiosink --speaker-codec=mp3
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=directsoundsink --speaker-codec=mp3
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=wasapisink --speaker-codec=mp3


**--speaker-codec=vorbis:**
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=autoaudiosink --speaker-codec=vorbis
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=directsoundsink --speaker-codec=vorbis
Xpra_cmd.exe -d sound,gstreamer --env=XPRA_SOUND_SINK=wasapisink --speaker-codec=vorbis

**Comment by totaam** [comment]
Thanks for the detailed investigation. This narrows it down quite a bit.

I believe this must be something to do with the pulseaudio server configuration.
Can you try to run `xpra showconfig | grep -i pactl` on your server?
You should be seeing:
```
pulseaudio-configure-commands           = 'pactl set-default-sink Xpra-Speaker', 'pactl set-default-source Xpra-Microphone.monitor'
```
These commands are meant to set the devices that applications should use, and it looks like they're picking the wrong one for some reason.

Perhaps `pactl` is not installed on the server?
Or maybe you can try running each command by hand from an xterm running in the xpra session and then start Firefox from there? Does that help?

**Comment by svargh** [comment]
Thanks for looking into the issue!
I did some pactl stuff and actually have no idea of pactl and pulseaudio, I hope the outputs help you:
- I checked in xpra terminal, pactl is installed on server.
- xpra showconfig | grep -i pactl on Server gives me the expected output:

```
xpra showconfig | grep -i pactl
pulseaudio-configure-commands           = 'pactl set-default-sink Xpra-Speaker', 'pactl set-default-source Xpra-Microphone.monitor'
```

- In the xpra session terminal: After executing the above commands, then starting Firefox, there is still no sound, also no silderchange sound in pavucontrol. There is also no error message:
[pactl_output.txt](https://github.com/Xpra-org/xpra/files/6137276/pactl_output.txt)
Executing pactl list retrieves some output, but imght be this will help debugging? 
Btw, actually sound works out-of-the box on that Ubuntu 20.04 LTS Desktop (Gnome)

- Before and after executing the pactl commands, firefox still showing the same Media Input and Output devices:
![xpra-issue-1_firefox_media_devices](https://user-images.githubusercontent.com/13705625/111079572-dc70c300-84fa-11eb-8f79-57bdbd5633da.png)

Still, Firefox sound & pavocontrol silder change sound  are only come on Xpra Microphone

Maybe, I should use another pulseaudio version?

**Comment by svargh** [comment]
I created a fresh new VMWare VM with Ubuntu Desktop 20.04 LTS (Minimal installation) and without soundcard and performed a fresh apt update&upgrade afterwards, and rebooted once.

Installed Xpra according to [https://github.com/Xpra-org/xpra/wiki/Download#step-by-step-example-for-installing-the-stable-repository-on-ubuntu-focal-fossa](url).

Also here, I get only on microphone, but not on speaker.

Just curious, which specific version and distribution can I used in a VM?

**Comment by svargh** [comment]
Just tried recent xpra beta software on the same Ubuntu 20.04 LTS Desktop,
but also here too, sound only on microphone.
- Client: https://xpra.org/beta/windows/Xpra-Python3-x86_64_Setup_4.2-r28593.exe
- Server: 
```
2021-03-14 22:00:34,028 xpra GTK3 X11 version 4.2 64-bit
2021-03-14 22:00:34,071 OpenGL is supported on display ':0'
2021-03-14 22:00:34,071  using 'llvmpipe (LLVM 11.0.0, 256 bits)' renderer

```

**Comment by totaam** [comment]
> Executing pactl list retrieves some output, but imght be this will help debugging?

The only thing that I see which may be relevant is:
```
window.x11.display = ":0"
```
Assuming that `:0` is the display your session is using, then the pulseaudio data is also correctly exported to the X11 display.

> Btw, actually sound works out-of-the box on that Ubuntu 20.04 LTS Desktop (Gnome)

Desktop environment sessions use the regular pulseaudio server.
Xpra sessions require a dedicated pulseaudio server per-session so that the audio does not leak into other desktops / sessions.

> Maybe, I should use another pulseaudio version?

No. Any pulseaudio version from the last 10 years should work.

I'll take a look when I have access to my virtual machines again, sometime next week.

**Comment by svargh** [comment]
Update:
1) I installed  this time a fresh ubuntu server image (ubuntu-20.04.1-live-server-amd64.iso)
with recent xpra beta ( xpra GTK3 X11 version 4.2 64-bit) and Client  https://xpra.org/beta/windows/Xpra-Python3-x86_64_Setup_4.2-r28593.exe

2)  Additionally, did apt install pavucontrol firefox.
**In Server version, sound works in xpra without problems, so closing this issue**

3) So in  my previous ubuntu-desktop installation, there is something interfering with xpra although I stopped gdm and X11 session on console and killed any open pulseaudio processes.

4) Server Details:
root@vmuser:/home/vmuser# uname -a
Linux vmuser 5.4.0-66-generic #74-Ubuntu SMP Wed Jan 27 22:54:38 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
root@vmuser:/home/vmuser#
xpra v4.2

5) If you have any ideas any directions, why it is not working in Ubuntu Desktop that would be great.
